File: Assignment3/MNIST_code/q5_code/loadMnist.m

function [train, test] = loadMnist()
  binary_digits = 1;
  nSamples = 30000; % Number of samples from the dataset
  % Load the training data
  X=(loadMNISTImages('train-images.idx3-ubyte'));
  X = X(:,1:nSamples);
  y=(loadMNISTLabels('train-labels.idx1-ubyte')');
  y = y(1:nSamples);

  %Digits to classify
  d1 = 0;
  d2 = 1;
  
  
  if (binary_digits)
    % Take only d1 and d2 digits
    X = [ X(:,y==d1), X(:,y==d2) ];
    y = [ y(y==d1), y(y==d2) ];
  end

  % Randomly shuffle the data
  I = randperm(length(y));
  y=y(I); % labels in range 1 to 10
  X=X(:,I);

  % We standardize the data so that each pixel will have roughly zero mean and unit variance.
  s=std(X,[],2);
  m=mean(X,2);
    %Create a matrix and calculate the standard deviation along each row.

  X=bsxfun(@minus, X, m);
  X=bsxfun(@rdivide, X, s+.1);

  % Place these in the training set
  train.X = X;
  train.y = y;

  % Load the testing data
  X=loadMNISTImages('t10k-images.idx3-ubyte');
  y=loadMNISTLabels('t10k-labels.idx1-ubyte')';
  if (binary_digits)
    % Take only the 0 and 1 digits
    X = [ X(:,y==d1), X(:,y==d2) ];
    y = [ y(y==d1), y(y==d2) ];
  end

  % Randomly shuffle the data
  I = randperm(length(y));
  y=y(I); % labels in range 1 to 10
  X=X(:,I);

  % Standardize using the same mean and scale as the training data.
  X=bsxfun(@minus, X, m);
  X=bsxfun(@rdivide, X, s+.1);

  % Place these in the testing set
  test.X=X;
  test.y=y;

File: Assignment3/loadMNIST.py
from mlxtend.data import mnist_data, loadlocal_mnist
import numpy as np
from sklearn.model_selection import train_test_split



# return X=[x1|...|Xm].trnspode()   R:mxn
# Y=[y1|...|ym].traranspose   R:mx1
def loadMnist(X,Y,d1=0,d2=1,nSamples = 30000):  # return [X,Y]
    # Dimensions x: 5000 x 784
    # Dimensions: y: 5000
    # X, Y = mnist_data()
    X_t = X.transpose()
    X = X[0:nSamples]
    Y = Y[0:nSamples]
    # Digits to classify
    # d1 = 0
    # d2 = 1
    # Take only d1 and d2 digits
    filtered_data = np.array([X[i] for i in range(0, X.shape[0]) if (Y[i] == d1 or Y[i] == d2)])  # (1000, 1)
    filtered_labels = np.array([[Y[i]] for i in range(0, X.shape[0]) if (Y[i] == d1 or Y[i] == d2)])  # (1000, 784)
    return (filtered_data.astype('float') / 255, filtered_labels)


# calculate the standard deviation and mean along each coloum.
def standardize_arrays(data):
    # rows std
    s = data.std(0)
    m = np.mean(data, axis=0)
    data = data - m
    data = data / (s + 1.0)
    return (data)


# Randomly shuffle the data,lables
def shuffle_array_the_same(a, b):
    assert len(a) == len(b)
    p = np.random.permutation(len(a))
    return a[p], b[p]


# return (X,Y)tset,(X,Y)train
# X=[x1|...|Xm].transpose()   R:nxm, n=num images, m=28^2 num of pixels in an image. x[0][4] pixel 4 in image 0
# Y=[y1|...|ym].transform()   R:mx1  Y[0] label of image 0 (X[0])
def random_shuffeled_Mnist(d1=0,d2=1):
    # (data, labels) = loadMnist()
    X_train, y_train = loadlocal_mnist(
        images_path='t10k-images.idx3-ubyte',
        labels_path='t10k-labels.idx1-ubyte')
    X_train, y_train=loadMnist(X_train, y_train,d1,d2)

    X_test, y_test = loadlocal_mnist(
        images_path='train-images.idx3-ubyte',
        labels_path='train-labels.idx1-ubyte')
    X_test, y_test = loadMnist(X_test, y_test,d1,d2)

    (train_data_shufled, train_labels_shufled) = shuffle_array_the_same(X_train, y_train)
    train_data_random_shuffeled = standardize_arrays(train_data_shufled)

    (test_data_shufled, test_labels_shufled) = shuffle_array_the_same(X_test, y_test)
    test_data_random_shuffeled = standardize_arrays(test_data_shufled)

    return train_data_random_shuffeled.transpose(), train_labels_shufled,test_data_random_shuffeled.transpose(), test_labels_shufled

